simd-support/simd-r5v.h

```diff
@@ -0,0 +1,473 @@
+#  define TYPEINT(name) __riscv_ ## name ## _u64m1
+#  define TYPEINT64(name) __riscv_ ## name ## _u64m1
+#  define TYPEMASK(name) __riscv_ ## name ## _f64m1_m
+#  define TYPEMERGEDMASK(name) __riscv_ ## name ## _f64m1_mu
+#  define TYPEINTERPRETF2U(name) __riscv_ ## name ## _f64m1_u64m1
+#  define TYPEINTERPRETU2F(name) __riscv_ ## name ## _u64m1_f64m1
+#endif /* FFTW_SINGLE */
+
+#if R5V_SIZE == 16384
+#define VL DS(128, 256)        /* SIMD complex vector length */
+#elif R5V_SIZE == 8192
+#define VL DS(64, 128)        /* SIMD complex vector length */
+#elif R5V_SIZE == 4096
+#define VL DS(32, 64)        /* SIMD complex vector length */
+#elif R5V_SIZE == 2048
+#define VL DS(16, 32)        /* SIMD complex vector length */
+#elif R5V_SIZE == 1024
+#define VL DS(8, 16)        /* SIMD complex vector length */
+#elif R5V_SIZE == 512
+#define VL DS(4, 8)        /* SIMD complex vector length */
+#elif R5V_SIZE == 256
+#define VL DS(2, 4)        /* SIMD complex vector length */
+#elif R5V_SIZE == 128
+#define VL DS(1, 2)        /* SIMD complex vector length */
+#else /* R5V_SIZE */
+#error "R5V_SIZE must be a power of 2 between 128 and 16384 bits"
+#endif /* R5V_SIZE */
+
+#define SIMD_VSTRIDE_OKA(x) ((x) == 2) 
+#define SIMD_STRIDE_OKPAIR SIMD_STRIDE_OK
+
+#if defined(__GNUC__) && !defined(__riscv_vector) /* sanity check */
+#error "compiling simd-r5v.h without RISC-V V support"
+#endif
+
+#include <riscv_vector.h>
+
+typedef DS(vfloat64m1_t, vfloat32m1_t) V;
+typedef DS(vuint64m1_t, vuint32m1_t) Vint;
+typedef DS(vuint64m1_t, vuint64m2_t) Vint64;
+typedef DS(vbool64_t, vbool32_t) Vmask;
+#define INT2MASK(imask) DS(__riscv_vmsne_vx_u64m1_b64, __riscv_vmsne_vx_u32m1_b32)(imask, 0, 2*VL)
+
+/* unused */
+/*
+#ifdef FFTW_SINGLE
+#warning "VLIT F32 perhaps should broadcast 64 bits"
+#endif
+static inline V VLIT(const R re, const R im) {
+  Vint idx = TYPEINT(vid_v)(2*VL); // (0, 1, 2, 3, ...)
+  Vint vone = TYPEINT(vmv_v_x)(1, 2*VL);
+  V vre = TYPE(vfmv_v_f)(re, 2*VL);
+  V vim = TYPE(vfmv_v_f)(im, 2*VL);
+  Vint mask = TYPEINT(vand_vv)(idx, vone, 2*VL); // (0, 1, 0, 1, 0, ...)
+  return  TYPE(vmerge_vvm)(vre, vim, INT2MASK(mask), 2*VL);
+}
+*/
+
+#define VLIT1(val) TYPE(vfmv_v_f)(val, 2*VL)
+#define LDK(x) x
+#define DVK(var, val) V var = VLIT1(val)
+#define VZERO VLIT1(DS(0.,0.f))
+/* unused */
+//#define VRONE VLIT(DS(1.,1.f),DS(0.,0.f))
+//#define VCI VLIT(DS(0.,0.f),DS(1.,1.f))
+//#define VCONEMI VLIT(DS(1.,1.f),DS(-1.,-1.f))
+//#define VONE  VLIT1(DS(1.,1.f))
+//#define VMINUSONE VLIT1(DS(-1.,-1.f))
+// generate (all 1, 0, all 1, 0, ...) to split real and imagine parts
+#define VPARTSPLIT TYPEINT(vsub_vx)(TYPEINT(vand_vx)(TYPEINT(vid_v)(2*VL), 1, 2*VL), 1, 2*VL)
+
+static inline V VDUPL(const V x) {
+	Vint idx = TYPEINT(vid_v)(2*VL); // (0, 1, 2, 3, ...)
+	//Vint vnotone = TYPEINT(vmv_v_x)(DS(~1ull,~1), 2*VL);
+	Vint hidx = TYPEINT(vand_vx)(idx, DS(~1ull,~1), 2*VL); // (0, 0, 2, 2, ...)
+	return TYPE(vrgather_vv)(x, hidx, 2*VL);
```

Comment:
The index of destination of vrgather_vv depends on the value in vector reg, which makes it relatively unpredictable for compiler/microarchitecture. May the vslide instruction have a more regular behaviour and better performance?